packages/dataDisplay/useSortableTable/src/useSortableTable.tsx

```diff
@@ -0,0 +1,99 @@
+import { useCallback, useEffect, useMemo, useState } from "react";
+
+type SortableTableDataType<V extends Object> = Map<keyof V, Array<V[keyof V]>>;
+
+const useSortableTable = <V extends Object>(data: V[]) => {
+    type K = keyof V;
+
+    const [convertedData, setConvertedData] = useState<SortableTableDataType<V> | undefined>(undefined);
+    const [sortableData, setSortableData] = useState<V[]>(data);
+
+    const keys = useMemo(()=> {
+        const objectKeys = data.reduce((acc, currentValue) => {
+            const currentValueKeys = Object.keys(currentValue) as K[];
+            
+            const newObjectKeys = currentValueKeys.filter((key) => !acc.includes(key));
+        
+            return [...acc, ...newObjectKeys];
+        }, [] as K[]);
+
+        return objectKeys;
+    }, [data]);
+
+    const _initalizeConvertedData = useCallback(() => {
+        const initTableData: SortableTableDataType<V> = new Map();
+
+        keys.forEach((key) => {
+            initTableData.set(key, []);
+        })
+    
+        data.forEach((rowData) => {
+            (Object.keys(rowData) as K[]).forEach((key) => {                
+                const targetMapArray = initTableData.get(key);
+
+                if (targetMapArray === undefined) {
+                    throw new Error("Can not find key in data");
+                }
+
+                targetMapArray.push(rowData[key]);
+            });
+        });
+    
+        setConvertedData(initTableData);
+    },[data, keys]);
+    
+    useEffect(() => {
+        _initalizeConvertedData();
+    },[]);
+
+    const sort = useCallback((key: K, compareFn: (a: any, b: any) => number) => {
+        if (keys.length === 0) {
+            return;
+        }
+
+        if (convertedData === undefined) {
+            throw new Error("tableData is not initialized");
+        }
+
+        const targetColumn = convertedData.get(key);
+
+        if (targetColumn === undefined) {
+            throw new Error("Can not find key in data");
+        }
+        
+        const sortedIndex = targetColumn.map((v, i) => ({v, i}))
+            .sort((a, b) => compareFn(a.v, b.v))
+            .map((sorted) => sorted.i);
+
+        const sortedData = sortedIndex.map((index) => 
+            keys.reduce((obj, key) => {
+                const targetMapArray = convertedData.get(key);
+                if (targetMapArray === undefined) {
+                    throw new Error("Can not find key in data");
+                }                
+                obj[key] = targetMapArray[index];
+                return obj;
+            }, {} as V)
```

Comment:
더 중요한 이야기를 하느라 따로 말씀 못드렸는데.. 저는 비대칭적인(교환법칙/결합법칙이 성립하지 않는) reduce를 싫어합니다. 😅 다음과 같이 작성하시면 더 가독성에서 좋을 것 같습니다.

```suggestion
            keys.map( (key) => {
                return { [key]: (convertedData.get(key)] || []) [index] } as V;
            }).reduce((objA, objB) => {
                let ret = Object.assign({}, objA);
                ret = Object.assign(ret, objB);
                
                return ret;
            }, {} as V)
```

Reply:
```tsx
keys.map( (key) => {
    const value = convertedData.get(key);

    if (!value) {
        throw new Error("Can not find key in data");
    }

    return { [key]: value[index] };
}).reduce((objA, objB) => {
    let ret = Object.assign({}, objA);
    ret = Object.assign(ret, objB);
    
    return ret;
}, {} as V)
```

value값이 undefined일 경우 예외처리만 추가해서 다음과 같이 수정하였습니다!